### PR TITLE
PLAT-101582: Cache item refs in VirtualList

### PR DESCRIPTION
**Pull request by ybsung**
### Checklist

* [x] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [ ] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [ ] At least one test case is included for this feature or bug fix
* [ ] Documentation was added or is not needed
* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)

The `querySelector` are used a lot in the VirtualList to access item DOM elements easily. We might avoid using DOM APIs directly.

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)

Cache item refs which are passed as the VirtualList's prop - `itemRefs` and use it when needed. If there is no `itemRefs` prop, then the item refs are not cached.

### Additional Considerations
[//]: # (How should the change be tested?)
[//]: # (Are there any outstanding questions?)
[//]: # (Were any side-effects caused by the change?)

### Links
[//]: # (Related issues, references)

PLAT-101582

### CODE REVIEW

File: packages/ui/VirtualList/VirtualListBasic.js [hunk 1]
```diff
@@ -1044,20 +1052,35 @@ class VirtualListBasic extends Component {
...
-			{itemProps, itemRenderer, getComponentProps} = this.props,
-			key = index % this.state.numOfItems,
-			componentProps = getComponentProps && getComponentProps(index) || {};
+			{itemRefs, itemProps, itemRenderer, getComponentProps} = this.props,
+			{numOfItems} = this.state,
+			key = index % numOfItems,
+			componentProps = getComponentProps && getComponentProps(index) || {},
+			itemContainerRef = (ref) => {
+				if (itemRefs) {
+					itemRefs.current[index % numOfItems] = (ref === null) ? null : ref;
```
Comment (MikyungKim) [inline packages/ui/VirtualList/VirtualListBasic.js]: Could we replace the `(ref === null) ? null : ref;` to 'ref ? ref : null` ?

Comment (MikyungKim) [reply packages/ui/VirtualList/VirtualListBasic.js]: Isn't `index % numOfItems` `key`?

Comment (0x64) [reply packages/ui/VirtualList/VirtualListBasic.js]: It looks the same as just `ref`.

File: packages/ui/VirtualList/VirtualListBasic.js [hunk 2]
```diff
@@ -1044,20 +1052,35 @@ class VirtualListBasic extends Component {
...
 		);
 	}
 
 	applyStyleToHideNode = (index) => {
-		const key = index % this.state.numOfItems;
-		this.cc[key] = <div key={key} style={{display: 'none'}} />;
+		const
+			{itemRefs} = this.props,
+			{numOfItems} = this.state,
+			key = index % this.state.numOfItems,
```
Comment (MikyungKim) [inline packages/ui/VirtualList/VirtualListBasic.js]: You mean `numOfItems` here?

File: packages/ui/VirtualList/VirtualListBasic.js [hunk 3]
```diff
@@ -1044,20 +1052,35 @@ class VirtualListBasic extends Component {
...
+			itemContainerRef = () => {
+				if (itemRefs) {
+					itemRefs.current[index % numOfItems] = null;
```
Comment (MikyungKim) [inline packages/ui/VirtualList/VirtualListBasic.js]: maybe `key`?

File: packages/ui/VirtualList/VirtualListBasic.js (hunk 1 above)
Comment (ybsung) [reply packages/ui/VirtualList/VirtualListBasic.js]: > Could we replace the `(ref === null) ? null : ref;` to `ref ? ref : null` ?

Is `ref || null` fine?

Comment (MikyungKim) [reply packages/ui/VirtualList/VirtualListBasic.js]: yes

Comment (ybsung) [reply packages/ui/VirtualList/VirtualListBasic.js]: > Isn't `index % numOfItems` `key`?

Yes.... I replaced that part with the key.

Comment (ybsung) [reply packages/ui/VirtualList/VirtualListBasic.js]: > It looks the same as just `ref`.

OK.

**Comment by ybsung** [comment]
I fixed all that @MikyungKim  and @0x64 mentioned.

File: packages/ui/Scroller/ScrollerBasic.js [hunk 4]
```diff
@@ -187,6 +187,7 @@ class ScrollerBasic extends Component {
 		delete rest.cbScrollTo;
 		delete rest.scrollContainerContainsDangerously;
 		delete rest.direction;
+		delete rest.itemRefs;
```
Comment (MikyungKim) [inline packages/ui/Scroller/ScrollerBasic.js]: Can we assign this prop only for VL?

File: packages/ui/VirtualList/VirtualListBasic.js [hunk 5]
```diff
@@ -1194,7 +1211,7 @@ class VirtualListBasic extends Component {
 		return (
 			<div className={containerClasses} data-webos-voice-focused={voiceFocused} data-webos-voice-group-label={voiceGroupLabel} data-webos-voice-disabled={voiceDisabled} ref={this.props.scrollContentRef} style={style}>
 				<div {...rest} className={contentClasses} ref={this.contentRef}>
-					{itemsRenderer({cc, itemContainerRef, primary})}
+					{itemsRenderer({cc, primary})}
```
Comment (MikyungKim) [inline packages/ui/VirtualList/VirtualListBasic.js]: I think we are referring to `itemContainerRef` in sandstone VL.
Please make a change in sandstone along with this.

Comment (ybsung) [reply packages/ui/VirtualList/VirtualListBasic.js]: Fix it in the sandstone PR and in the agate PR.

File: packages/ui/VirtualList/VirtualListBasic.js [hunk 6]
```diff
@@ -993,7 +1000,7 @@ class VirtualListBasic extends Component {
 
 	// For individually sized item
 	adjustItemPositionWithItemSize () {
-		if (this.itemContainerRef.current) {
+		if (this.contentRef.current) {
```
Comment (MikyungKim) [inline packages/ui/VirtualList/VirtualListBasic.js]: I think this one is not identical to the old condition.
Maybe `cc.length` would be better.

**Pull request merged**